### PR TITLE
List all certficates in the UI

### PR DESCRIPTION
**Pull request by stephdl**
Actually only certificate created by the UI of manually by the action set-certificate are listed inside the UI

previously only route starting by : certificate-sub.domain.com@file and custom certificate were listed inside the UI (with also custom certificate)

now we list all routes if the LE certificate has been requested and we verify if it has been obtained or not.

for that we do a curl on the traefik api and we push the name (`*@file`) to get-certificate

for backward compatibility for example the UI or other action like  set-certificate or delete-certificate I still offer to listen after fqdn (previous behavior)

### CODE REVIEW

**Comment by Amygos** [comment]
A couple of things:
1. I don't see where the attribute `type` is vauled  as `custom`
2. We need to add some test coverage for the new attribute:
  * https://github.com/NethServer/ns8-traefik/blob/5a473e9e0d0d2f23f017a9b7c4589a416df68a7b/tests/20_traefik_certificates_api.robot#L25
  * https://github.com/NethServer/ns8-traefik/blob/5a473e9e0d0d2f23f017a9b7c4589a416df68a7b/tests/20_traefik_certificates_api.robot#L34

**Comment by stephdl** [comment]
hello @Amygos the custom is inherited from https://github.com/NethServer/ns8-traefik/blob/main/imageroot/pypkg/custom_certificate_manager.py#L30C6-L30C6

I did not change that part

for the test I look after

**Comment by stephdl** [comment]
it should be good @Amygos

File: imageroot/actions/list-certificates/validate-output.json [hunk 1]
```diff
@@ -1,22 +1,64 @@
...
-	    "oneOf":[{
-		    "type": "object",
-		    "title": "A certificate expanded"
-	    },
-	    {
-		    "type": "string",
-		    "format": "hostname",
-		    "title": "A fully qualified domain name"
-	    }]
+        "anyOf": [
```
[the remark this answers is not in the document]
Comment (Amygos) [reply imageroot/actions/list-certificates/validate-output.json:32]: Update. there is a bug in the json scheme validator that allows a mixed array.

**Comment by Amygos** [comment]
> hello @Amygos the custom is inherited from https://github.com/NethServer/ns8-traefik/blob/main/imageroot/pypkg/custom_certificate_manager.py#L30C6-L30C6

I see, there is a problem with the custom certificate listing, the `expand-list` attribute is not honored.
> ```
> [root@webserver ~]# api-cli run list-certificates --agent module/traefik1 | jq
> Warning: using user "cluster" credentials from the environment
> [
>   "tata.sdl.nethserver.net",
>   "toto.sdl.nethserver.net",
>   "nextcloud.sdl.nethserver.net",
>   "webserver2.sdl.nethserver.net",
>   {
>     "fqdn": "stephane",
>     "type": "custom",
>     "obtained": true
>   }
> ]
> ```
The above output is not valid, but a bug in the json schema validator let the mixed output to pass the validation.

**Comment by Amygos** [comment]
Please update also the json schema of  the `get-certificate` action
https://github.com/NethServer/ns8-traefik/blob/main/imageroot/actions/get-certificate/validate-output.json

File: imageroot/actions/list-certificates/validate-output.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephdl) [reply imageroot/actions/list-certificates/validate-output.json:32]: can you explain please, this is what I output, like you can see I could have a list of fqdn and custom certi if they exist

```
[root@webserver ~]# api-cli run list-certificates --agent module/traefik1 --data '{"expand_list": true}' | jq
Warning: using user "cluster" credentials from the environment
[
  {
    "fqdn": "nextcloud2.sdl.nethserver.netCopy",
    "type": "internal",
    "obtained": false
  },
  {
    "fqdn": "toto.sdl.nethserver.net",
    "type": "internal",
    "obtained": true
  },
  {
    "fqdn": "webserver.sdl.nethserver.net",
    "type": "internal",
    "obtained": true
  },
  {
    "fqdn": "webserver.sdl.nethserver.net",
    "type": "route",
    "obtained": true
  },
  {
    "fqdn": "stephane",
    "type": "custom",
    "obtained": true
  }
]
[root@webserver ~]# api-cli run list-certificates --agent module/traefik1 | jq
Warning: using user "cluster" credentials from the environment
[
  "nextcloud2.sdl.nethserver.netCopy",
  "toto.sdl.nethserver.net",
  "webserver.sdl.nethserver.net",
  "webserver.sdl.nethserver.net",
  {
    "fqdn": "stephane",
    "type": "custom",
    "obtained": true
  }
]
```

[the remark this answers is not in the document]
Comment (Amygos) [reply imageroot/actions/list-certificates/validate-output.json:32]: As explained in https://github.com/NethServer/ns8-traefik/pull/45#issuecomment-1798633225, the correct output of `api-cli run list-certificates --agent module/traefik1` should be:
```json
[
  "nextcloud2.sdl.nethserver.netCopy",
  "toto.sdl.nethserver.net",
  "webserver.sdl.nethserver.net",
  "webserver.sdl.nethserver.net",
  "stephane"
]
```
> The output of `list-certificate` must be a list of strings or objects ( the same object returned by the `get-certificate` action) if the `expand-list` is `true`.

Let me rephrase better: must be a list of strings **or** ( mutually exclusive)  a list of objects. not a list of objects and strings.

[the remark this answers is not in the document]
Comment (stephdl) [reply imageroot/actions/list-certificates/validate-output.json:32]: agree but it was like that, no change on my side, do I need to change it , it will be a new behavior

cc @DavidePrincipi

[the remark this answers is not in the document]
Comment (stephdl) [reply imageroot/actions/list-certificates/validate-output.json:32]: change have been introduced by https://github.com/NethServer/ns8-traefik/pull/34/files

**Comment by stephdl** [comment]
> Please update also the json schema of the `get-certificate` action https://github.com/NethServer/ns8-traefik/blob/main/imageroot/actions/get-certificate/validate-output.json

done

File: imageroot/actions/list-certificates/validate-output.json [hunk 2]
```diff
@@ -1,22 +1,72 @@
...
+                    "type",
+                    "obtained"
+                ]
+            }
+        },
+        {
+            "minItems": 0,
+            "maxItems": 0
+        }
+    ]
```
Comment (Amygos) [inline imageroot/actions/list-certificates/validate-output.json:71]: There is no need to use `minItems`/maxItems` if we use `anyOf`
```suggestion
    "anyOf": [
        {
            "items": {
                "type": "string",
                "format": "idn-hostname"
            }
        },
        {
            "items": {
                "type": "object",
                "properties": {
                    "fqdn": {
                        "type": "string",
                        "format": "idn-hostname"
                    },
                    "type": {
                        "type": "string",
                        "enum": [
                            "internal",
                            "custom",
                            "route"
                        ]
                    },
                    "obtained": {
                        "type": "boolean"
                    }
                },
                "required": [
                    "fqdn",
                    "type",
                    "obtained"
                ]
            }
        }
    ]
```

File: imageroot/actions/list-certificates/validate-output.json [hunk 3]
```diff
@@ -1,22 +1,72 @@
 {
     "$schema": "http://json-schema.org/draft-07/schema#",
     "title": "list-certificates output",
-    "$id": "http://schema.nethserver.org/samba/list-certificates-output.json",
+    "$id": "http://schema.nethserver.org/traefik/list-certificates-output.json",
     "description": "Return a list of requested certificates fqdn",
     "examples": [
-	    ["example.com"],
-	    []
```
Comment (Amygos) [inline imageroot/actions/list-certificates/validate-output.json:8]: The empty array is a valid output

**Review by Amygos — CHANGES_REQUESTED** [review]

**Comment by Amygos** [comment]
@stephdl There is also a new bug, a custom certificate is listed twice:
```
[root@rl1 ~]# api-cli run module/traefik1/list-certificates  -d '{"expand_list": true}' | jq
Warning: using user "cluster" credentials from the environment
[
  {
    "fqdn": "example.net",
    "type": "internal",
    "obtained": false
  },
  {
    "fqdn": "test.it",
    "type": "custom",
    "obtained": true
  },
  {
    "fqdn": "test.it",
    "type": "custom",
    "obtained": true
  }
]
```

**Review by Amygos — APPROVED** [review]
